## Gradient Descent – Concept and Intuition

Gradient Descent is an optimization algorithm used to minimize a loss function by iteratively updating model parameters in a direction that reduces prediction error. It is the core learning mechanism behind many machine learning models and neural networks.

---

## Why Gradient Descent Is Required

Most machine learning models define a loss function that measures the difference between predicted and actual values. For complex models, finding the optimal parameters analytically is either infeasible or computationally expensive. Gradient Descent provides an efficient iterative approach to approximate the optimal solution.

---

## Core Intuition

The algorithm works by computing the gradient (slope) of the loss function with respect to the model parameters and updating the parameters in the opposite direction of the gradient.

- A steep slope indicates a large error and results in larger updates.

- A shallow slope indicates smaller updates as the model approaches convergence.

This process continues until the loss stabilizes or reaches a minimum.

---

Mathematical Formulation

The parameter update rule is defined as:

θ = θ - α⋅∇J(θ)

Where:

- θ represents the model parameters (weights)

- α is the learning rate

- ∇J(θ) is the gradient of the loss function with respect to the parameters

---

## Learning Rate (α)

The learning rate controls how large each update step is.

- A large learning rate may cause the model to diverge.

- A small learning rate may slow down convergence.

Choosing an appropriate learning rate is critical for stable and efficient training.

---

## Gradient Descent in Neural Networks

Neural network loss functions are differentiable, allowing gradients to be efficiently computed using backpropagation. Gradient Descent then updates weights layer by layer, enabling the network to learn complex patterns from data.

---

## Practical Considerations

Feature scaling improves convergence speed

Monitoring loss curves helps diagnose training issues

Advanced optimizers such as Momentum, RMSProp, and Adam improve convergence stability

---

## Summary

Gradient Descent is a fundamental optimization technique that enables models to learn from data by iteratively minimizing error through gradient-based parameter updates.

## Gradient Descent using Tensorflow

In [ ]:
import numpy as np
import pandas as pd
import tensorflow as tf
from tensorflow import keras
import matplotlib.pyplot as plt
%matplotlib inline

In [ ]:
df = pd.read_csv("insurance_data.csv")
df.head()

,age,affordibility,bought_insurance
0,22,1,0
1,25,0,0
2,47,1,1
3,52,0,0
4,46,1,1


### Split Train and Test set

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df[['age','affordibility']],df.bought_insurance,test_size=0.2, random_state = 25)

In [ ]:
X_train_scaled = X_train.copy()
X_train_scaled['age'] = X_train_scaled['age'] / 100

X_test_scaled = X_test.copy()
X_test_scaled['age'] = X_test_scaled['age'] / 100

In [ ]:
model = keras.Sequential([
    keras.layers.Dense(1, input_shape = (2,), activation='sigmoid', kernel_initializer = 'ones', bias_initializer='zeros')
])

model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(X_train_scaled, y_train, epochs=500)


Epoch 1/500


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 692ms/step - accuracy: 0.5000 - loss: 0.7113
Epoch 2/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 139ms/step - accuracy: 0.5000 - loss: 0.7110
Epoch 3/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 71ms/step - accuracy: 0.5000 - loss: 0.7106
Epoch 4/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 131ms/step - accuracy: 0.5000 - loss: 0.7102
Epoch 5/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 64ms/step - accuracy: 0.5000 - loss: 0.7098
Epoch 6/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 142ms/step - accuracy: 0.5000 - loss: 0.7094
Epoch 7/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 139ms/step - accuracy: 0.5000 - loss: 0.7091
Epoch 8/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 69ms/step - accuracy: 0.5000 - loss: 0.7087
Epoch 9/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 61ms/step - accuracy: 0.5000 - loss: 0.7083
Epoch 10/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 65ms/step - accuracy: 0.5000 - loss: 0.7079
Epoch 11/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 66ms/step - accuracy: 0.5000 - loss: 0.7076
Epoch 12/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 67ms/step - accuracy: 0.5000 - loss: 0.7

In [ ]:
model.evaluate(X_test_scaled,y_test)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 151ms/step - accuracy: 0.6667 - loss: 0.6210


[0.621041476726532, 0.6666666865348816]

In [ ]:
model.predict(X_test_scaled)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 103ms/step


array([[0.66075516],
       [0.6084721 ],
       [0.45149705],
       [0.6268543 ],
       [0.66423655],
       [0.68474233]], dtype=float32)

In [ ]:
y_test

,bought_insurance
2,1
10,0
21,0
11,0
14,1
9,1


In [ ]:
coef, intercept = model.get_weights()

In [ ]:
coef, intercept

(array([[0.7785071 ],
        [0.69779867]], dtype=float32),
 array([-0.39703575], dtype=float32))

## Gradient Descent for scratch

### Sigmoid Function

In [ ]:
def sigmoid(x):
  return 1 / (1 + np.exp(-x))

### Prediction Function

In [ ]:
def prediction_function(age, affordibility):
  weighted_sum = coef[0]*age + coef[1]*affordibility + intercept
  return sigmoid(weighted_sum)

### Loss Function

In [ ]:
def log_loss(y_true, y_predicted):
  epsilon = 1e-15
  y_predicted_new = [max(i,epsilon) for i in y_predicted]
  y_predicted_new = [min(i,1-epsilon) for i in y_predicted_new]
  y_predicted_new = np.array(y_predicted_new)

  return -np.mean(y_true * np.log(y_predicted_new) + (1-y_true) * np.log(1-y_predicted_new))

## Gradient Descent

In [ ]:
def gradient_descent(age, affordibility, y_true, epochs, loss_threshold):
  w1 = w2 = 1
  bias = 0
  rate = 0.5
  n = len(age)

  for i in range(epochs):
    weighted_sum = (w1 * age) + (w2 * affordibility) + bias
    y_predicted = sigmoid(weighted_sum)
    loss = log_loss(y_true, y_predicted)

    w1d = (1/n) * np.dot(np.transpose(age),(y_predicted - y_true))
    w2d = (1/n) * np.dot(np.transpose(affordibility),(y_predicted - y_true))

    bias_d = np.mean(y_predicted - y_true)

    w1 = w1 - rate * w1d
    w2 = w2 - rate * w2d
    bias = bias - rate * bias_d

    print(f'Epochs:{i}, w1:{w1}, w2:{w2}, bias:{bias}, loss:{loss}')

    if loss <= loss_threshold:
      break

    return w1,w2,bias


In [ ]:
gradient_descent(X_train_scaled['age'], X_train_scaled['affordibility'],y_train, 1000, 0.6364)

Epochs:0, w1:0.974907633470177, w2:0.948348125394529, bias:-0.11341867736368583, loss:0.7113403233723417


(np.float64(0.974907633470177),
 np.float64(0.948348125394529),
 np.float64(-0.11341867736368583))

In [ ]:
coef, intercept

(array([[0.7785071 ],
        [0.69779867]], dtype=float32),
 array([-0.39703575], dtype=float32))

## What the Coefficients Represent

Each coefficient corresponds to the weight assigned to an input feature.

### Weight 1 = 0.7785
- Indicates that the first feature has a strong positive influence on the model’s prediction.

### Weight 2 = 0.6978
- Indicates that the second feature also contributes positively to the output.

A positive coefficient means that as the feature value increases, the predicted output increases, assuming other features remain constant.

---

## What the Intercept (Bias) Represents

The intercept value of −0.3970 represents the model’s baseline prediction when all input feature values are zero.

This bias term allows the model to better fit the data by shifting the prediction line (or decision boundary) upward or downward.

---

Model Equation

The learned linear model can be expressed as:

y^ = 0.7785 * x1 + 0.6978 * x2 - 0.3970

---

## Summary

The learned coefficients and intercept indicate that the Gradient Descent implementation successfully optimized the model parameters, resulting in a stable and interpretable linear relationship between input features and the predicted output.